### ISSUE TITLE
CAS validation issue with synapse

### ISSUE THREAD

**Issue by tyxieblub** [matrix-org/synapse#2639]
Hello,

### Description

It seems there is a problem with how CAS validation is handled in [login.py#L420](https://github.com/matrix-org/synapse/blob/632baf799ea876d2346e934dce90d6a24cb92e37/synapse/rest/client/v1/login.py#L420).

### Steps to reproduce

- Install and configure CAS 5.1.x (I didn't try with earlier ones)
- Install matrix-synapse
- Configure matrix-synapse to use CAS to connect with `service_url: https://matrix.domain`
- Try to login!

You'll see it doesn't work and if I'm not mistaken it is because in [login.py#L397](https://github.com/matrix-org/synapse/blob/632baf799ea876d2346e934dce90d6a24cb92e37/synapse/rest/client/v1/login.py#L397) we ask for a ticket with a service in the form of `https://matrix.domain/_matrix/client/api/v1/login/cas/ticket?redirectUrl=***`.
But when we validate the ticket, we only send `cas_service_url` which lacks the `/_matrix/client/api/v1/login/cas/ticket?redirectUrl=***` part so CAS fails the validation saying the ticket was used for another service than the one it was created for.

I don't know if earlier versions were more lenient on validation, but I got it to work by replacing line 420 by this line:
```python
"service": self.cas_service_url + request.uri[:request.uri.find("&ticket")]
```

But I'm sure there's a better way to do it...

I believe #2404 may be related to this problem.

### Version information

- **Homeserver**: local homeserver

If not matrix.org:
- **Version**: matrix-synapse 0.24.1-1    
- **Install method**: Debian repository
- **Platform**: VM on Debian9

**Comment by johappel** [comment]
same issue

**Comment by mijutu** [comment]
The problem still exists in matrix-synapse-py3 0.99.2+stretch1

[matrix-synapse-0.99.2-CAS-fix.diff.gz](https://github.com/matrix-org/synapse/files/2949876/matrix-synapse-0.99.2-CAS-fix.diff.gz)

The service URL given to CAS is not supposed to change as long as synapse configuration stays the same, so we can have CasRedirectServlet store the url somewhere and later get it back in CasTicketServlet

**Comment by quorak** [comment]
This ticket is s old. I came across the same problem. So how are others using cas. It seems impossible to me, or can there be a less strict validation on the cas side?

we should implement the patch though! thanks @mijutu

**Comment by mikifus** [comment]
I had the same problem. I can confirm the patch works.

**Comment by SeaLife** [comment]
I dont feel creating a new issue will suit this cause i think this is still related to this one...

So we wanna move to Riot.IM but selfhosted and we want to use our CAS server for authentication.

The CAS Server we'r using is only capable of doing XML (like CAS 1.0)...
Configuration seems to be working on the first glance but configured the Service like this:

```
cas_config:
   enabled: true
   server_url: "https://authy.example.de/cas"
   service_url: "https://matrix.example.de"
#   required_attributes:
#       name: username
```

We'r getting redirected to our CAS server and redirected to:

```
https://matrix.example.de/_matrix/client/r0/login/cas/ticket?redirectUrl=http%3A%2F%2Fchat.example.de%2F%3Fhomeserver%3Dhttps%253A%252F%252Fmatrix.example.de%26identityServer%3D&ticket=ST-2987327
```

But getting there will raise this:

```
{
    "errcode": "M_UNKNOWN",
    "error": "Internal server error"
}
```

For some reason Synapse is not even trying to resolve the ticket. The CAS server we'r using is logging this and the ticket keeps getting expired without getting called:

```
2019-10-24 22:00:35.274  INFO 1 --- [io-8080-exec-10] CASTicketHandling                        : Issued new Ticket for TicketData(service=https://matrix.example.de/_matrix/client/r0/login/cas/ticket?redirectUrl=http%3A%2F%2Fchat.example.de%2F%3Fhomeserver%3Dhttps%253A%252F%252Fmatrix.example.de%26identityServer%3D, expiresAt=Thu Oct 24 22:00:50 GMT 2019) with Ticket ID ST-2987327
2019-10-24 22:00:57.209  WARN 1 --- [   scheduling-1] CASTicketHandling                        : Ticket ST-2987327 expired.
```

Also the DEBUG log is not giving me any more information but i dont know if the DEBUG log is in the docker output. We'r using the Docker-Image `matrixdotorg/synapse:latest` 

Regards

**Comment by mijutu** [comment]
The old patch didn't apply, so I made a new one.

[matrix-synapse-1.5.1-CAS-fix.diff.gz](https://github.com/matrix-org/synapse/files/3819383/matrix-synapse-1.5.1-CAS-fix.diff.gz)

**Comment by richvdh** [comment]
so somebody opened a pull request with @mijitu's patch at https://github.com/matrix-org/synapse/pull/5044; however I had questions about it which were never answered so we had to close the PR.

@mijutu: perhaps you could open a PR so that we can work on landing your changes?

**Comment by mijutu** [comment]
We had a pull request and there richvdh commented that the solution I used was wrong. I think he is correct. If two CAS logins happen at the same time, then users might end up at a wrong url.

Probability of two simultaneous CAS logins is small because CAS is used only at the initial login. Browsers remember access tokens and CAS is rarely used.

Homeserver owners might be able to use my patch for months or years without having issues, but if it's wrong, then it's wrong and shouldn't be merged.

**Comment by mijutu** [comment]
[matrix-synapse-1.7.2-CAS-fix.diff.gz](https://github.com/matrix-org/synapse/files/4012323/matrix-synapse-1.7.2-CAS-fix.diff.gz)

Here is a better fix for the problem. CasTicketServlet generates the service parameter from synapse config and redirectUrl parameter of the request.

**Comment by richvdh** [comment]
@mijutu please could you open a PR with your patch so that we can read and discuss it?

**Comment by richvdh** [comment]
fixed by #6634. Thanks @Naugrimm!
